pkg/restore/prioritize_group_version.go

```diff
@@ -0,0 +1,337 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restore
+
+import (
+	"context"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/version"
+
+	velerov1api "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+	"github.com/vmware-tanzu/velero/pkg/archive"
+	"github.com/vmware-tanzu/velero/pkg/client"
+)
+
+// ChosenGroupVersion is the API Group version that was selected to restore
+// from potentially multiple backed up version enabled by the feature flag
+// APIGroupVersionsFeatureFlag
+type ChosenGroupVersion struct {
+	Group   string
+	Version string
+	Dir     string
+}
+
+// chooseAPIVersionsToRestore will choose a version to restore based on a user-
+// provided config map prioritization or our version prioritization.
+func (ctx *restoreContext) chooseAPIVersionsToRestore() error {
+	sourceGVs, targetGVs, userGVs, err := ctx.gatherSourceTargetUserGroupVersions()
+	if err != nil {
+		return err
+	}
+
+OUTER:
+	for rg, sg := range sourceGVs {
+		// Default to the source preferred version if no other common version
+		// can be found.
+		cgv := ChosenGroupVersion{
+			Group:   sg.Name,
+			Version: sg.PreferredVersion.Version,
+			Dir:     sg.PreferredVersion.Version + velerov1api.PreferredVersionDir,
+		}
+
+		tg := findAPIGroup(targetGVs, sg.Name)
+		if len(tg.Versions) == 0 {
+			ctx.chosenGrpVersToRestore[rg] = cgv
+			ctx.log.Debugf("Chose %s/%s API group version to restore", cgv.Group, cgv.Version)
+			continue
+		}
+
+		// Priority 0: User Priority Version
+		if userGVs != nil {
+			uv := findSupportedUserVersion(userGVs[rg].Versions, tg.Versions, sg.Versions)
+			if uv != "" {
+				cgv.Version = uv
+				cgv.Dir = uv
+
+				if uv == sg.PreferredVersion.Version {
+					cgv.Dir += velerov1api.PreferredVersionDir
+				}
+
+				ctx.chosenGrpVersToRestore[rg] = cgv
+				ctx.log.Debugf("APIGroupVersionsFeatureFlag Priority 0: User defined API group version %s chosen for %s", uv, rg)
+				continue
+			}
+
+			ctx.log.Infof("Cannot find user defined version in both the cluster and backup cluster. Ignoring version %s for %s", uv, rg)
+		}
+
+		// Priority 1: Target Cluster Preferred Version
+		if versionsContain(sg.Versions, tg.PreferredVersion.Version) {
+			cgv.Version = tg.PreferredVersion.Version
+			cgv.Dir = tg.PreferredVersion.Version
+
+			if tg.PreferredVersion.Version == sg.PreferredVersion.Version {
+				cgv.Dir += velerov1api.PreferredVersionDir
+			}
+
+			ctx.chosenGrpVersToRestore[rg] = cgv
+			ctx.log.Debugf(
```

Comment:
I like that the previous case (priority 0) has an `Info` log message that it was unsuccessful. Could you add `Info` lines for when cases 1-3 fail to match too, so the user has more visibility into what the code decided without having to switch on debug logs? Switching on debug logs isn't a huge deal, but it does mean stopping the Velero deployment and re-running the backup job.

Reply:
Good idea--will do.